homeassistant/components/legrandinone.py

```diff
@@ -0,0 +1,473 @@
+        connection = create_iobl_connection(
+            port=port,
+            host=host,
+            event_callback=event_callback,
+            disconnect_callback=reconnect,
+            loop=hass.loop,
+            ignore=None
+        )
+
+        try:
+            with async_timeout.timeout(CONNECTION_TIMEOUT,
+                                       loop=hass.loop):
+                transport, protocol = await connection
+
+        except (serial.serialutil.SerialException, ConnectionRefusedError,
+                TimeoutError, OSError, asyncio.TimeoutError) as exc:
+            reconnect_interval = config[DOMAIN][CONF_RECONNECT_INTERVAL]
+            _LOGGER.exception(
+                "Error connecting to iobl, reconnecting in %s",
+                reconnect_interval)
+
+            hass.loop.call_later(reconnect_interval, reconnect, exc)
+            return
+
+        # There is a valid connection to a IOBL device now so
+        # mark entities as available
+        async_dispatcher_send(hass, SIGNAL_AVAILABILITY, True, protocol)
+
+        # Bind protocol to command class to allow entities to send commands
+        hass.data[IOBL_PROTOCOL_HANDLE] = protocol
+
+        # handle shutdown of IOBL asyncio transport
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP,
+                                   lambda x: transport.close())
+
+        _LOGGER.info('Connected to Iobl')
+
+    hass.async_create_task(connect())
+    return True
+
+
+class LegrandInOneDevice(Entity):
+    """Representation of an iobl device.
+
+    Contains the common logic for Iobl entities.
+    """
+
+    platform = None
+    _state = None
+    _available = True
+    legrand_id = None
+    iobl_mode = None
+    iobl_media = None
+    iobl_type = None
+    iobl_unit = None
+    _protocol = None
+
+    def __init__(self, device_id, protocol, initial_event=None, name=None,
+                 iobl_media='plc', iobl_comm_mode='unicast'):
+        """Initialize the device."""
+        self._initial_event = initial_event
+        self.legrand_id = device_id
+        self.iobl_mode = iobl_comm_mode
+        self.iobl_media = iobl_media
+        self._protocol = protocol
+
+        if name:
+            self._name = name
+        else:
+            self._name = device_id
+
+    @callback
+    def handle_event_callback(self, event):
+        """Handle incoming event for device type."""
+        # Call platform specific event handler
+        self._handle_event(event)
+
+        # Propagate changes through ha
+        self.async_schedule_update_ha_state()
+
+    def _handle_event(self, event):
+        """Platform specific event handler."""
+        raise NotImplementedError()
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def name(self):
+        """Return a name for the device."""
+        return self._name
+
+    @property
+    def is_on(self):
```

Comment:
Only common properties should be implemented in the common entity class. Eg cover entitites don't support this property.